components/concurrency_manager/src/lib.rs

```diff
@@ -52,9 +75,61 @@ impl ConcurrencyManager {
 
     /// Updates max_ts with the given new_ts. It has no effect if
     /// max_ts >= new_ts or new_ts is TimeStamp::max().
-    pub fn update_max_ts(&self, new_ts: TimeStamp) {
-        if new_ts != TimeStamp::max() {
-            self.max_ts.fetch_max(new_ts.into_inner(), Ordering::SeqCst);
+    ///
+    /// To avoid invalid ts breaking the invariants, the new_ts should be
+    /// less than or equal to the max_ts_limit.
+    ///
+    /// # Returns
+    /// - Ok(()): If the update is successful or has no effect
+    /// - Err(limit): If new_ts is greater than the max_ts_limit, returns the
+    ///   current limit value
+    pub fn update_max_ts(
+        &self,
+        new_ts: TimeStamp,
+        source: Option<String>,
```

Comment:
Why not receiving `&str`? Passing `"...".to_owned()` looks potentially introducing additional overhead.

Reply:
Ah... I see. Perhaps we can have a enum type like `UpdateMaxSource`, and format it into string only when necessary.

Reply:
🤔
How about: 
```rust
fn update_max_ts(&self, new_ts: TimeStamp, source: impl Display)
```
So the enum type can be defined outside the concurrency manager module(?)

Reply:
And they it doesn't need to be a single type (e.g., CDC can have its source representation in `cdc` module, while txn related requests can have an enum defined in the `tikv` module)

Reply:
Using the generic arguments seems a good idea.
Will `format_args` be enough so we don't have to define types everywhere?